log/log.go

```diff
@@ -0,0 +1,114 @@
+package log
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+)
+
+type loggingLevel int
+
+// Log messages can have one of 3 possible states.
+const (
+	DEBUG loggingLevel = 1 << iota
+	INFO
+	CRITICAL
+)
+
+// List of fields that should not be overwritten by fields specified when logging
+// a message
+var restrictedFields = []string{"level", "time", "message"}
+
+// Creates a new logging instance
+func New(out io.Writer, level loggingLevel, fields map[string]interface{}) *Logger {
+	return &Logger{
+		out:           out,
+		defaultFields: fields,
+		level:         level,
+	}
+}
+
+type Logger struct {
+	// Create a lock when writing to `out` so output is not intermingled
+	mu  sync.Mutex
+	out io.Writer
+	// List of default fields that will be serialized
+	// along with the message and written to `out`
+	defaultFields map[string]interface{}
+	level         loggingLevel
+}
+
+// Log a debug message as long as the logger's debug level is set to at least Debug
+func (l *Logger) Debug(message string, fields map[string]interface{}) {
```

Comment:
I'm curious have you considered doing something like `logrus`.... ie. declare a `Tags` type like they have `Fields`... and accept `a ...interface{}` as message (var args are nice)...
Using something like `WithFields` is also nice, as we might not always have fields...

Honestly, I'm not a fan of doing our own logger, just using an existing library to have loggers available in various context where they are useful.

-----
Note: I'm not opposed to using `logrus`, I haven't really looked at it... but it seems to support child loggers...

Ie. it would be possible to do something like:
```go
log := logrus.New();

log.WithFields(Fields{"tag": "value"}).Debug("Hello", "world", 5)
log.Debug("Hello", "world", 5)

// Create task log for TaskContext (this is for system messages, not the livelog that is uploaded to queue)
taskLog := log.WithFields(Fields{"taskId": "...", "runId": 1})
taskLog.WithFields(Fields{"pulling": "img", "time": 100}).Debug("Pulling an image: ", img)
taskLog.Debug("Pulled an image: ", img)
```

If we think that `logrus` exposes too many methods.. And we want to do some hiding... 
We can just define an interface, containing only the methods we want plugins and engines to use.
For example I see no need to expose `log.Reader()`...